Replicate medusa bucket secrets to each Cass DC

rzvoncek opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

**What this PR does**:

This PR adds a bucket secret reconciliation process. If Medusa is configured globally (meaning the cluster's medusa config features a `medusaConfigurationRef`, then this process will make sure the bucket key secrets are properly replicated into the namespaces (and clusters) of each of the DCs. It will also alter the k8ssandra cluster to use the copied secret, instead of the original one.

This PR does not implement the actual replication, it only makes a copy of the secret and places labels on it. We make a copy so that we can clean up after ourselves (the original secret is user-created and we should not touch it). The labels are then picked up by the secret reconciliation process which actually copies them over.

**Which issue(s) this PR fixes**:
Fixes #1159.

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

burmanm commented on controllers/k8ssandra/medusa_reconciler.go:
```diff
@@ -229,7 +235,7 @@ func mergeStorageProperties(
 	if err := remoteClient.Get(ctx, configKey, storageProperties); err != nil {
 		logger.Error(err, fmt.Sprintf("failed to get MedusaConfiguration %s", configKey))
-		return result.Error(err)
+		return result.RequeueSoon(r.DefaultDelay)
```
question: Why the change? The only difference I can see is the change of delay when reconcile happens (error will do a reconcile also - but with a backoff timer to prevent flooding the server which this doesn't).

burmanm replied:
issue: This error handling is still incorrect, in case of error, err should be returned for proper requeue process (especially in this case when it can be Kubernetes API server overload or similar and we don't want to make it worse)

burmanm commented on controllers/k8ssandra/medusa_reconciler.go:
```diff
@@ -247,3 +253,63 @@ func mergeStorageProperties(
+	medusaConfigKey := types.NamespacedName{Namespace: medusaConfigNamespace, Name: medusaConfigName}
+	if err := remoteClient.Get(ctx, medusaConfigKey, medusaConfig); err != nil {
+		logger.Error(err, "failed to get MedusaConfiguration", "key", medusaConfigKey)
+		return result.RequeueSoon(r.DefaultDelay)
```
question: Why not result.Error(err) ? Now we don't get any proper processing of the error. Only if you wish to stop reconcile should this be result.Done() or something similar, but for requeue, always let the controller-runtime do that one for errors.

rzvoncek replied:
I had no good reason for this. Saw some weird behaviour and though requeueing is going to help. Now I know better.
Reverted to returing the error.

burmanm commented on controllers/k8ssandra/medusa_reconciler.go:
```diff
@@ -247,3 +253,63 @@ func mergeStorageProperties(
+	err := remoteClient.Get(ctx, bucketSecretKey, bucketSecret)
+	if err != nil {
+		logger.Error(err, "failed to get bucket secret", "key", bucketSecretKey)
+		return result.RequeueSoon(r.DefaultDelay)
```
question: Same here, result.Error(err) is better.

rzvoncek replied:
Also, changed to return the error.

burmanm commented on controllers/k8ssandra/medusa_reconciler.go:
```diff
@@ -247,3 +253,63 @@ func mergeStorageProperties(
+		newSecretRef := corev1.LocalObjectReference{Name: clusterBucketSecret.Name}
+		kc.Spec.Medusa.StorageProperties.StorageSecretRef = newSecretRef
+		if err = remoteClient.Update(ctx, kc); err != nil {
+			logger.Error(err, "failed to update K8ssandraCluster, re-queueing")
```
.

burmanm commented on controllers/k8ssandra/medusa_reconciler.go:
```diff
@@ -247,3 +253,63 @@ func mergeStorageProperties(
+		kc.Spec.Medusa.StorageProperties.StorageSecretRef = newSecretRef
+		if err = remoteClient.Update(ctx, kc); err != nil {
+			logger.Error(err, "failed to update K8ssandraCluster, re-queueing")
+			return result.RequeueSoon(r.DefaultDelay)
```
.

burmanm commented:
issue: I see no changes in the envtests, would you be able to add those as well? That provides so much faster response to changes so it helps other developers more than the e2e tests (which realistically, no one is running on their machine and wait until GHA throws something and they're not as easily debuggable).

rzvoncek commented:
Addressed the issue. Added env tests and updated e2e tests. 
While doing this, also changed when we actually do the bucket secret reconciliation. Now it's not happening for each DC, but only once, at the place where we deal with all Medusa-related secrets.

burmanm commented on controllers/k8ssandra/medusa_reconciler.go:
```diff
@@ -119,7 +120,7 @@ func (r *K8ssandraClusterReconciler) reconcileMedusa(
 		if !ready {
 			logger.Info("Medusa standalone deployment is not ready yet")
-			return result.RequeueSoon(r.DefaultDelay)
+			return result.Error(err)
```
There's no err necessarily here, the method can return a `false, nil` in which case err is nil here.

rzvoncek replied:
Great spot, fixed.

rzvoncek commented:
Did another round of improvements. The return error vs. requeue was indeed a problem. All green now (aside from codecov).

adejanovski commented on docs/content/en/tasks/backup-restore/_index.md:
```diff
@@ -106,9 +106,11 @@ stringData:
 The file should always specify `credentials` as shown in the example above; in that section, provide the expected format and credential values that are expected by Medusa for the chosen storage backend. For more, refer to the [Medusa documentation](https://github.com/thelastpickle/cassandra-medusa/blob/master/docs/Installation.md) to know which file format should used for each supported storage b ...
 
-A successful deployment should inject a new init container named `medusa-restore` and a new container named `medusa` in the Cassandra StatefulSet pods.  
+If using a shared Medusa configuration (see below), this secret can be created in the same namespace as the `MedusaConfiguration` object. The K8ssandra operator will then make sure the secret is replicated to the namespaces hosting the Cassandra clusters.
```
```suggestion
If using a shared Medusa configuration (see below), this secret has to be created in the same namespace as the `MedusaConfiguration` object. k8ssandra-operator will then make sure the secret is replicated to the namespaces hosting the Cassandra clusters.
```

adejanovski reviewed (approved):
Brilliant!
Everything worked perfectly in my tests, with both namespace and cluster scope.

I put a suggestion change for some wording in the docs.
